Lucene memory segment input regression caused by JDK 23

**ES version**:
main / lucene_snapshot

**JDK**: 
openjdk version "23" 2024-09-17
OpenJDK Runtime Environment (build 23+37-2369)
OpenJDK 64-Bit Server VM (build 23+37-2369, mixed mode, sharing)

**OS**:
* bare metal, Ubuntu 18.04 LTS, kernel 5.4.0-65-generic x86_64, 8 cores, 32 GiB RAM - reproducible occasionally
* GCP, Ubuntu 22.04 LTS, kernel 6.5.0-1025-gcp x86_64, 8 cores, 8 GiB RAM - reproducible every time
* GCP, Ubuntu 22.04 LTS, kernel 5.15.0-1025-gcp x86_64, 8 cores, 8 GiB RAM - reproducible every time

**Bug**:
Lucene memory segment input slows down with JDK 23. This applies to both Lucene 9.11.1 as well as most recent 9.12.0 snapshot. The regression was noticed in nightly benchmarks when switched temporarily to JDK 23 on September 4th (rolled back on September 6th).

For example nightly benchmark for [Nested](https://github.com/elastic/rally-tracks/tree/master/nested) Rally track regressed as follows for `nested-date-histo` query ([src](https://github.com/elastic/rally-tracks/blob/54d3433dcca0f239ddedd3d26b8cbefed9eeb44f/nested/operations/default.json#L45-L66)) latency:

([dashboard](https://elasticsearch-benchmarks.elastic.co/#tracks/nested/nightly/default/90d))

![Image](https://github.com/user-attachments/assets/5c0ada1a-ceef-4768-bffe-f0180cbba2fe)

Wall clock profiles collected when running this query continuously show that ES spends way more time in methods such as `org.apache.lucene.store.MemorySegmentIndexInput.readLong()`.

JDK 22:

![Image](https://github.com/user-attachments/assets/c664ca3e-8f20-4353-aa34-9bdc08fc62da)

![Image](https://github.com/user-attachments/assets/6dcf21c3-fb3b-4393-9292-7d02f863edd9)

JDK 23:

![Image](https://github.com/user-attachments/assets/1703395d-6fdf-41b2-bdb2-b842bcf5ea15)

![Image](https://github.com/user-attachments/assets/111ce6d4-a4ba-4d12-974f-915141c903b2)

[wall-clock-profiles.zip](https://github.com/user-attachments/files/17030845/wall-clock-profiles.zip)

commits between jdk-23+30 and jdk-23+31:

```
6720685abdc1c91f5083255782810cdb2d7d72f4 (tag: jdk-23+31) 8335966: Remove incorrect problem listing of java/lang/instrument/NativeMethodPrefixAgent.java in ProblemList-Virtual.txt
4aab58be4aa1bbe848084189b467d0c586a4ffbb 8335637: Add explicit non-null return value expectations to Object.toString()
5b9ecb178667e373a834b241b8f8c8a13cef3b4d 8333886: Explicitly specify that asSlice and reinterpret return a memory segment backed by the same region of memory.
70ad622bc23fc5a1808466193fd6c7ea2f178ec9 8335124: com/sun/management/ThreadMXBean/ThreadCpuTimeArray.java failed with CPU time out of expected range
ca37a482cce9ad2bd57221f7ae34ac8246cd1843 8334757: AssertionError: Missing type variable in where clause
ae10055b2cbd00734d41feef5251b52883151abb 8335824: Test gc/arguments/TestMinInitialErgonomics.java is timing out
4e5232097902b7d30f9bd6e21f3936d29462375f 8334433: jshell.exe runs an executable test.exe on startup
2f60d3684843e42c2c4c3bf628ac6d62dcdfcadb 8335475: ClassBuilder incorrectly calculates max_locals in some cases
b415b981397395eac298dfc841d1a9d012f62501 8334297: (so) java/nio/channels/SocketChannel/OpenLeak.java should not depend on SecurityManager
90d5b5b4c497ac99d0e2ade689b6459fffea3e2a 8324089: Fix typo in the manual page for "jcmd" (man jcmd)
653c481d71b3c5c6bc1e867d88a560495557876c 8335479: JFR: Missing documentation for -XX:StartFlightRecording
10b28babe53821fcdeef3a1aa0712feb7cd67529 8322812: Manpage for jcmd is missing JFR.view command
d383365ea4196cd5f40de217547392b820c4ad01 8334890: Missing unconditional cross modifying fence in nmethod entry barriers
b6d0ead93fbfff92418defc13d9c4c7b2f95a72a (HEAD, tag: jdk-23+30) 8335221: Some C2 intrinsics incorrectly assume that type argument is compile-time constant
```

--- a/distribution/src/config/jvm.options
+++ b/distribution/src/config/jvm.options
@@ -58,7 +58,7 @@
 # result in less optimal vector performance
 20-:--add-modules=jdk.incubator.vector
 
-# Required to workaround perform issue in JDK 23, https://github.com/elastic/elasticsearch/issues/113030
+# Required to workaround performance issue in JDK 23, https://github.com/elastic/elasticsearch/issues/113030
 23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.setAsTypeCache
 23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.asTypeUncached
 

--- a/distribution/src/config/jvm.options
+++ b/distribution/src/config/jvm.options
@@ -59,8 +59,8 @@
 20-:--add-modules=jdk.incubator.vector
 
 # Required to workaround perform issue in JDK 23, https://github.com/elastic/elasticsearch/issues/113030
-23-23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.setAsTypeCache
-23-23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.asTypeUncached
+23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.setAsTypeCache
+23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.asTypeUncached
 
 ## heap dumps
 

--- a/distribution/src/config/jvm.options
+++ b/distribution/src/config/jvm.options
@@ -58,6 +58,10 @@
 # result in less optimal vector performance
 20-:--add-modules=jdk.incubator.vector
 
+# Required to workaround perform issue in JDK 23, https://github.com/elastic/elasticsearch/issues/113030
+23-23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.setAsTypeCache
+23-23:-XX:CompileCommand=dontinline,java/lang/invoke/MethodHandle.asTypeUncached
+
 ## heap dumps
 
 # generate a heap dump when an allocation from the Java heap fails; heap dumps

